File: benchmarks/mandelbrot_simd/python/mandelbrot_simd.py
import sys

import numpy as np

MAX_ITER = 1000


def escape_iters_single(cr, ci):
    zr = zi = 0.0
    i = 0
    while i < MAX_ITER:
        zr2, zi2 = zr * zr, zi * zi
        if zr2 + zi2 > 4.0:
            break
        zi = 2.0 * zr * zi + ci
        zr = zr2 - zi2 + cr
        i += 1
    return i


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 800

    assert escape_iters_single(0.0, 0.0) == MAX_ITER, "self-check failed: origin should never escape"
    assert escape_iters_single(2.0, 2.0) < MAX_ITER, "self-check failed: far point should escape quickly"

    py, px = np.mgrid[0:n, 0:n]
    ci_grid = -1.5 + 3.0 * py / n
    cr_grid = -2.0 + 3.0 * px / n

    zr = np.zeros((n, n))
    zi = np.zeros((n, n))
    iters = np.zeros((n, n), dtype=np.int64)
    active = np.ones((n, n), dtype=bool)

    for _ in range(MAX_ITER):
        zr2 = zr * zr
        zi2 = zi * zi
        still = (zr2 + zi2 <= 4.0) & active
        if not still.any():
            break
        new_zi = 2.0 * zr * zi + ci_grid
        new_zr = zr2 - zi2 + cr_grid
        zr = np.where(still, new_zr, zr)
        zi = np.where(still, new_zi, zi)
        iters = np.where(still, iters + 1, iters)
        active = still

    count = int(np.sum(iters == MAX_ITER))
    print(count)


if __name__ == "__main__":
    main()

File: benchmarks/mandelbrot_simd/java/Mandelbrot_simd.java
import jdk.incubator.vector.DoubleVector;
import jdk.incubator.vector.LongVector;
import jdk.incubator.vector.VectorMask;
import jdk.incubator.vector.VectorOperators;
import jdk.incubator.vector.VectorSpecies;

public class Mandelbrot_simd {
    static final int MAX_ITER = 1000;
    static final VectorSpecies<Double> SPECIES_D = DoubleVector.SPECIES_256;
    static final VectorSpecies<Long> SPECIES_L = LongVector.SPECIES_256;
    static final int W = SPECIES_D.length();

    static long[] escapeCountsRow(double[] cr, double ci) {
        DoubleVector zr = DoubleVector.zero(SPECIES_D);
        DoubleVector zi = DoubleVector.zero(SPECIES_D);
        DoubleVector crVec = DoubleVector.fromArray(SPECIES_D, cr, 0);
        DoubleVector ciVec = DoubleVector.broadcast(SPECIES_D, ci);
        LongVector iters = LongVector.zero(SPECIES_L);
        VectorMask<Long> active = SPECIES_L.maskAll(true);

        for (int i = 0; i < MAX_ITER; i++) {
            DoubleVector zr2 = zr.mul(zr);
            DoubleVector zi2 = zi.mul(zi);
            DoubleVector mag2 = zr2.add(zi2);
            VectorMask<Double> cmpD = mag2.compare(VectorOperators.LE, 4.0);
            VectorMask<Long> still = cmpD.cast(SPECIES_L).and(active);

            if (!still.anyTrue()) break;

            DoubleVector newZi = zr.mul(zi).mul(2.0).add(ciVec);
            DoubleVector newZr = zr2.sub(zi2).add(crVec);
            VectorMask<Double> stillD = still.cast(SPECIES_D);
            zr = zr.blend(newZr, stillD);
            zi = zi.blend(newZi, stillD);

            iters = iters.add(1, still);
            active = still;
        }

        long[] out = new long[W];
        iters.intoArray(out, 0);
        return out;
    }

    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 800;

        long[] originIters = escapeCountsRow(new double[]{0.0, 0.0, 0.0, 0.0}, 0.0);
        for (long it : originIters) {
            if (it != MAX_ITER) {
                System.err.println("self-check failed: origin should never escape");
                System.exit(1);
            }
        }

        long[] farIters = escapeCountsRow(new double[]{2.0, 2.0, 2.0, 2.0}, 2.0);
        for (long it : farIters) {
            if (it >= MAX_ITER) {
                System.err.println("self-check failed: far point should escape quickly");
                System.exit(1);
            }
        }

        int groups = n / W;
        long count = 0;
        for (int py = 0; py < n; py++) {
            double ci = -1.5 + 3.0 * py / n;
            for (int gx = 0; gx < groups; gx++) {
                double[] cr = new double[W];
                for (int lane = 0; lane < W; lane++) {
                    int px = gx * W + lane;
                    cr[lane] = -2.0 + 3.0 * px / n;
                }
                long[] iters = escapeCountsRow(cr, ci);
                for (long it : iters) {
                    if (it == MAX_ITER) count++;
                }
            }
        }

        System.out.println(count);
    }
}
